packages/@romejs/compiler/lint/rules/react/noArrayIndexKey.ts

```diff
@@ -0,0 +1,219 @@
+	return path.findAncestry((path) => {
+		if (
+			path.node.type === "JSExpressionStatement" &&
+			path.node.expression.type === "JSCallExpression" &&
+			path.node.expression.callee.type === "JSMemberExpression" &&
+			path.node.expression.callee.property.value.type === "JSIdentifier"
+		) {
+			// Children
+			if (
+				path.node.expression.callee.object.type === "JSReferenceIdentifier" &&
+				path.node.expression.callee.object.name === "Children" &&
+				path.node.expression.callee.property.value.type === "JSIdentifier" &&
+				(path.node.expression.callee.property.value.name === "map" ||
+				path.node.expression.callee.property.value.name === "forEach")
+			) {
+				return true;
+			}
+
+			// React.Children
+			if (
+				path.node.expression.callee.object.type === "JSMemberExpression" &&
+				path.node.expression.callee.object.object.type ===
+				"JSReferenceIdentifier" &&
+				path.node.expression.callee.object.object.name === "React" &&
+				path.node.expression.callee.object.property.type ===
+				"JSStaticMemberProperty" &&
+				path.node.expression.callee.object.property.value.type ===
+				"JSIdentifier" &&
+				path.node.expression.callee.object.property.value.name === "Children" &&
+				(path.node.expression.callee.property.value.name === "map" ||
+				path.node.expression.callee.property.value.name === "forEach")
+			) {
+				return true;
+			}
+		}
+		return false;
+	});
+}
+
+function hasArrayMethod(path: Path) {
+	return path.findAncestry((path) => {
+		if (
+			path.node.type === "JSExpressionStatement" &&
+			path.node.expression.type === "JSCallExpression" &&
+			path.node.expression.callee.type === "JSMemberExpression" &&
+			path.node.expression.callee.property.type === "JSStaticMemberProperty" &&
+			path.node.expression.callee.property.value.type === "JSIdentifier"
+		) {
+			return /[map|forEach|filter|some|every|find|findIndex|reduce|reduceRight|]/.test(
+				path.node.expression.callee.property.value.name,
+			);
+		}
+		return false;
+	});
+}
+
+function hasArrayIndexKey(keyValue: string, path: Path) {
+	if (
+		path.node.type === "JSExpressionStatement" &&
+		path.node.expression.type === "JSCallExpression" &&
+		path.node.expression.arguments.length &&
+		isFunctionNode(
+			path.node.expression.arguments[path.node.expression.arguments.length - 1],
+		)
+	) {
+		const params = (path.node.expression.arguments[path.node.expression.arguments.length -
+		1] as JSFunctionExpression | JSArrowFunctionExpression).head.params;
+		if (
+			params.length > 1 &&
+			(params[params.length - 1] as JSBindingIdentifier).name === keyValue
+		) {
+			return true;
+		}
+	}
+
+	if (
+		path &&
+		isFunctionNode(path.node) &&
+		path.node.head.params.length &&
+		path.node.head.params[path.node.head.params.length - 1].type ===
+		"JSBindingIdentifier"
+	) {
+		const indexName = (path.node.head.params[path.node.head.params.length - 1] as JSBindingIdentifier).name;
+
+		return indexName === keyValue;
+	}
+
+	return false;
+}
```

Comment:
We can simplify this a lot by requiring just a `node` and then some normalization.

```suggestion
function hasArrayIndexKey(keyValue: string, node: AnyNode) {
	if (
		node.type === "JSExpressionStatement" &&
		node.expression.type === "JSCallExpression" &&
		node.expression.arguments.length > 0
	) {
		const lastArg = node.expression.arguments[node.expression.arguments.length - 1];

		if (isFunctionNode(lastArg)) {
			node = lastArg;
		}
	}

	if (
		isFunctionNode(node) &&
		node.head.params.length > 0
	) {
		const lastParam = node.head.params[node.head.params.length - 1];
		return lastParam.name === keyValue;
	}

	return false;
}
```